### PR TITLE
Changes to MBR within FlashLFQ

### PR DESCRIPTION
**Pull request by Alexander-Sol**
Major changes to MBR within FlashLFQ.

### Peak Decoys
When performing MBR, we attempt to locate a decoy peak for each acceptor peak. Acceptor peaks are located by predicting a retention time and looking for the theoretical peptide mass at that retention time. Decoy peaks are located by choosing a "random" retention time and looking for the theoretical peptide mass at the random retention time. 

### PEP Calculation for MBR Peak
All MBR peaks are fed into a binary classifier (similar to the one used within MetaMorpheus) and semi-supervised learning is performed to calculate a PEP for each peak.

### MBR Q-Values
For each file, Q-Values are calculated for each MBR peak by ranking them in PEP order and performing target-decoy competition

---

This pull request increases both the accuracy and precision of FlashLFQ. This was tested by analyzing a spike-in dataset, originally published by [Shen et al.](https://www.pnas.org/doi/10.1073/pnas.1800541115?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed). In this dataset, e. coli peptides were spiked into human peptides at 5 different ratios (1x, 1.5x, 2x, 2.5x, 3x). There were four replicates for each condition/ratio.

These files were analyzed with the previous version of FlashLFQ (with MBR and normalization enabled, denoted as v1.0 below) and the new version (with MBR and normalization enabled, 1% PIP FDR, 0.2% Donor FDR, denoted as PIP-ECHO below) . These results show that the new version of FlashLFQ has increased quantitative accuracy ( the average fold change is closer to the true fold change) and precision (smaller variance and CV).

Spike-In Comparison

| Comparison   | v1.0 Avg FC | PIP-ECHO Avg. FC | Empirical FC |
| :---------------- | ------: | ----: |------: | 
| 1x vs 1.5x         | 0.484 | 0.497 | 0.585 | 
| 1x vs 2x           | 0.857 | 0.871 | 1 | 
|1x vs 2.5x         | 1.16 | 1.16 | 1.32 | 
| 1x vs 3x           |  1.48 | 1.48 | 1.58 | 

| Comparison   |  v1.0 CV | PIP-ECHO CV | 
| :---------------- | ----: |------: | 
| 1x vs 1.5x         |  0.854 | 0.619 | 
| 1x vs 2x           |  0.544 | 0.405 | 
|1x vs 2.5x         | 0.436 | 0.324 | 
| 1x vs 3x           |  0.370 | 0.276 | 

---
This PR does not change the behavior of FlashLFQ if MBR is not enabled. The image below shows that there is a perfect correspondence between the values reported by the previous version of FlashLFQ and the update if MBR is disabled. Changes only occur when MBR is used. Data used was the 2-file FlashLFQ vignette data.

image

### CODE REVIEW

File: mzLib/MassSpectrometry/MassSpectrometry.csproj [hunk 1]
```diff
@@ -13,6 +13,7 @@
   <ItemGroup>
     <PackageReference Include="CsvHelper" Version="32.0.3" />
     <PackageReference Include="MathNet.Numerics" Version="5.0.0" />
+    <PackageReference Include="Plotly.NET.CSharp" Version="0.12.0" />
```
Comment (nbollis) [inline mzLib/MassSpectrometry/MassSpectrometry.csproj:16]: Is this intended?

File: mzLib/mzLib.nuspec
Comment (nbollis) [inline mzLib/mzLib.nuspec:1]: Indenting and version number are off

File: mzLib/FlashLFQ/IndexedMassSpectralPeak.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            return Mz.GetHashCode();
+            return HashCode.Combine(Mz.GetHashCode(), ZeroBasedMs1ScanIndex.GetHashCode());
```
Comment (nbollis) [inline mzLib/FlashLFQ/IndexedMassSpectralPeak.cs:32]: Minor point, mostly of curiosity: do you need to hash an index or is the index unique enough?

File: mzLib/FlashLFQ/ChromatographicPeak.cs [hunk 3]
```diff
@@ -40,27 +44,85 @@ public ChromatographicPeak(Identification id, bool isMbrPeak, SpectraFileInfo fi
...
+                && ApexRetentionTime == peak.ApexRetentionTime;
         }
 
+
         public IsotopicEnvelope Apex { get; private set; }
         public List<Identification> Identifications { get; private set; }
         public int NumChargeStatesObserved { get; private set; }
         public int NumIdentificationsByBaseSeq { get; private set; }
         public int NumIdentificationsByFullSeq { get; private set; }
         public double MassError { get; private set; }
```
Comment (nbollis) [inline mzLib/FlashLFQ/ChromatographicPeak.cs:61]: Most of your new properties are not directly covered by tests. Is this worrisome?

File: mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+using System.Threading.Tasks;
+using Omics;
+using Microsoft.FSharp.Collections;
+using System.Collections;
+using System.Security.Policy;
+using System.Text.RegularExpressions;
+
+namespace FlashLFQ.PEP
+{
+    public class DonorGroup : IEnumerable<ChromatographicPeak>
```
Comment (nbollis) [inline mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs:19]: Should break this class out into its own file, especially if it is public

File: mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+        }
+
+        public double BestTargetMbrScore => TargetAcceptors.Count == 0 ? 0 : TargetAcceptors.Max(acceptor => acceptor.MbrScore);
+
+        public IEnumerator<ChromatographicPeak> GetEnumerator()
+        {
+            return TargetAcceptors.Concat(DecoyAcceptors).GetEnumerator();
+        }
+
+        private IEnumerator GetEnumerator1()
```
Comment (nbollis) [inline mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs:39]: What is going on with GetEnumerator1?

File: mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+                    while (Math.Abs(targetSurplus) > 1 & !stuck & innerIterations < 25)
+                    {
+                        innerIterations++;
+                        swapped = false;
+                        foreach (int index in groupsOfIndices[i].OrderByDescending(idx => idx))
+                        {
+                            if (targetSurplus > 0)
+                            {
+                                if (donors[index].TargetAcceptors.Count(peak => peak.MbrScore > PipScoreCutoff) > 1)  // if we have a positive number of targets
+                                {
```
Comment (nbollis) [inline mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs:297]: There are a bunch of conditionals and indices being thrown around here without test coverage for a crucial part of the code. Could this be adjusted or tested?

File: mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+            {
+                logLossAverage = logLoss.Average();
+            }
+
+            if ((logLossReduction != null) && (logLossReduction.Any()))
+            {
+                logLossReductionAverage = logLossReduction.Average();
+            }
+
+            StringBuilder s = new StringBuilder();
```
Comment (nbollis) [inline mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs:627]: If a gradient boosted decision tree with semi-supervised learning uses the final tree, does it make sense to take the average of them all for output?

**Review by nbollis — CHANGES_REQUESTED** [review]
Mostly minor points. 

Several classes have inconsequential one line changes. 
PsmFromTsv, MsFraggerPeptide, MsFraggerPeptideFile

Several of your massive methods are untested in what appears to be the most dangerous part of the code, this worries me.

File: mzLib/FlashLFQ/IndexedMassSpectralPeak.cs (hunk 2 above)
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/IndexedMassSpectralPeak.cs:32]: The index probably doesn't need to be converted, you're right. I actually think you can pass in a double (mz) and an int (index) and get a hashcode out, elminating 2/3 hash operations

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 8]
```diff
@@ -106,8 +126,10 @@ public FlashLfqEngine(
             int mcmcBurninSteps = 1000,
             bool useSharedPeptidesForProteinQuant = false,
             bool pairedSamples = false,
-            List<string> peptideSequencesToUse = null,
-            int? randomSeed = null)
+            int? randomSeed = null,
+            char donorCriterion = 'S',
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:130]: enum?

File: mzLib/FlashLFQ/SpectraFileInfo.cs [hunk 9]
```diff
@@ -39,5 +39,9 @@ public override int GetHashCode()
         {
             return FullFilePathWithExtension.GetHashCode();
         }
+        public override string ToString()
+        {
+            return FilenameWithoutExtension;
```
Comment (trishorts) [inline mzLib/FlashLFQ/SpectraFileInfo.cs:44]: don't know the context here from looking at files changed but keeping the extension helps with .raw/.mzML confusion.

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 10]
```diff
@@ -545,18 +588,97 @@ private RetentionTimeCalibDataPoint[] GetRtCalSpline(SpectraFileInfo donor, Spec
...
+            switch (DonorCriterion)
+            {
+                case 'S': // Select best peak by the PSM score
+                    bestPeak = peaks.MaxBy(peak => peak.Identifications.Max(id => id.PsmScore));
+                    if (bestPeak.Identifications.First().PsmScore > 0)
+                        break;
+                    else // if every ID has a score of zero, let it fall through to the default case
+                        goto default;
+                case 'N': // Select peak with the most neighboring peaks
+                    int maxPeaks = 0;
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:656]: unit test case N

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 11]
```diff
@@ -855,36 +1073,95 @@ private void QuantifyMatchBetweenRunsPeaks(SpectraFileInfo idAcceptorFile)
...
-                            best.MergeFeatureWith(peak, Integrate);
+                            break;
+                        }
+                    }
+                    if (best == null) continue;
 
-                            peaksToRemoveFromHypotheses.Add(peak);
+                    // merge peaks with different charge states
+                    if (peakHypotheses.Count > 0)
+                    {
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:1104]: this if should probably get unit tests

File: mzLib/FlashLFQ/FlashLFQResults.cs [hunk 12]
```diff
@@ -59,6 +58,13 @@ public FlashLfqResults(List<SpectraFileInfo> spectraFiles, List<Identification>
             }
         }
 
+        public void ReNormalizeResults(bool integrate = false, int maxThreads = 10, bool useSharedPeptides = false)
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLFQResults.cs:61]: unit test?

File: mzLib/FlashLFQ/ChromatographicPeak.cs [hunk 13]
```diff
@@ -16,20 +16,23 @@ public class ChromatographicPeak
         public int ScanCount => IsotopicEnvelopes.Count;
         public double SplitRT;
         public readonly bool IsMbrPeak;
```
Comment (trishorts) [inline mzLib/FlashLFQ/ChromatographicPeak.cs:18]: We are really bulking up ChromatographicPeak. Is there anything to the idea of grouping a bunch of stuff into another class ChromatographicPeakMetrics? Just asking.

File: mzLib/FlashLFQ/FlashLFQ.csproj [hunk 14]
```diff
@@ -13,6 +13,8 @@
   <ItemGroup>
     <PackageReference Include="CsvHelper" Version="32.0.3" />
     <PackageReference Include="MathNet.Numerics" Version="5.0.0" />
+    <PackageReference Include="Microsoft.ML" Version="2.0.0" />
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLFQ.csproj:16]: This is up to 3.0.1 now. I think we updated MM. Maybe this should be updated too?

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 15]
```diff
@@ -489,52 +525,59 @@ private void QuantifyMs2IdentifiedPeptides(SpectraFileInfo fileInfo)
             _results.Peaks[fileInfo].AddRange(chromatographicPeaks.ToList());
         }
 
+        #region PeptideIdentityPropagation
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:528]: MBR....

**Review by trishorts — COMMENTED** [review]
looking good. Having a little old v. new data in the description of the pr would be good.

File: mzLib/FlashLFQ/ChromatographicPeak.cs (hunk 13 above)
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/ChromatographicPeak.cs:18]: Honestly that would be a better way to do that. Create something like and FdrInfo type object. I hope to implement that change in a future update

File: mzLib/FlashLFQ/FlashLFQResults.cs [hunk 16]
```diff
@@ -59,6 +58,17 @@ public FlashLfqResults(List<SpectraFileInfo> spectraFiles, List<Identification>
...
+            foreach(var peak in Peaks.SelectMany(p => p.Value))
+            {
+                peak.CalculateIntensityForThisFeature(integrate);
+            }
+            new IntensityNormalizationEngine(this, integrate, silent: true, maxThreads).NormalizeResults();
+            CalculatePeptideResults(quantifyAmbiguousPeptides: false);
+            CalculateProteinResultsMedianPolish(useSharedPeptides: useSharedPeptides);
+        }
+
         public void MergeResultsWith(FlashLfqResults mergeFrom)
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLFQResults.cs:72]: add method description

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 17]
```diff
@@ -855,36 +1080,95 @@ private void QuantifyMatchBetweenRunsPeaks(SpectraFileInfo idAcceptorFile)
...
                         {
-                            best.MergeFeatureWith(peak, Integrate);
+                            if (!peakHypotheses.Any())
+                            {
+                                best = null;
+                                break;
+                            }
+                            best = peakHypotheses.First();
+                            peakHypotheses.Remove(best);
+                        }
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:1101]: were these sections not tested previously ?

File: mzLib/FlashLFQ/FlashLfqEngine.cs [hunk 18]
```diff
@@ -855,36 +1080,95 @@ private void QuantifyMatchBetweenRunsPeaks(SpectraFileInfo idAcceptorFile)
...
                 }
-
-                _results.Peaks[idAcceptorFile].Add(best);
             }
 
-
             RunErrorChecking(idAcceptorFile);
         }
 
+        private void AddPeakToConcurrentDict(ConcurrentDictionary<string, ConcurrentDictionary<IsotopicEnvelope, List<ChromatographicPeak>>> matchBetweenRunsIdentifiedPeaks,
```
Comment (trishorts) [inline mzLib/FlashLFQ/FlashLfqEngine.cs:1140]: method description

File: mzLib/FlashLFQ/PEP/DonorGroup.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FlashLFQ.PEP
+{
+    public class DonorGroup : IEnumerable<ChromatographicPeak>
```
Comment (trishorts) [inline mzLib/FlashLFQ/PEP/DonorGroup.cs:10]: What is a "DonorGroup"? Please comment

File: mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs [hunk 20]
```diff
@@ -0,0 +1,625 @@
...
+        }
+
+        /// <summary>
+        /// Equalizes partitions used for cross-validation. The goal is to have the same number of targets and decoys in each partition
+        /// </summary>
+        /// <param name="donors"> List of all DonorGroups to be classified</param>
+        /// <param name="groupsOfIndices">An array of lists. Each list contains the indices of donor groups for a given partition </param>
+        /// <param name="scoreCutoff"> The MBR Score cutoff that determines which MBR target peaks will be used as positive training examples</param>
+        /// /// <param name="numGroups">Number of groups used for cross-validation, default = 3 </param>
+        public static void EqualizeDonorGroupIndices(List<DonorGroup> donors, List<int>[] groupsOfIndices, double scoreCutoff, int numGroups = 3)
```
Comment (trishorts) [inline mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs:234]: does this need to be so complicated

File: mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs [hunk 21]
```diff
@@ -0,0 +1,625 @@
...
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Takes in a list of donor groups and a list of indices for each group, and swaps two groups of indices
+        /// Updates the targetSurplus and decoySurplus variables
+        /// Updates the swappedDonors hash set to keep track of which donors have been swapped
```
Comment (trishorts) [inline mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs:374]: why?

File: mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs [hunk 22]
```diff
@@ -0,0 +1,625 @@
...
+                        ChromatographicPeakDataList.AddRange(localChromatographicPeakDataList);
+                    }
+                });
+
+            ChromatographicPeakData[] pda = ChromatographicPeakDataList.ToArray();
+
+            return pda.AsEnumerable();
+        }
+
+        public IEnumerable<ChromatographicPeakData> CreateChromatographicPeakDataIteration(List<DonorGroup> donors, List<int> donorIndices, int maxThreads)
```
Comment (trishorts) [inline mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs:452]: method comment

File: mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs [hunk 23]
```diff
@@ -0,0 +1,625 @@
...
+                donors[donorIndexA].DecoyAcceptors.Count);
+
+            groupsOfIndices[groupsOfIndicesIndexA].Add(donorIndexB);
+            groupsOfIndices[groupsOfIndicesIndexA].Remove(donorIndexA);
+            groupsOfIndices[groupsOfIndicesIndexB].Add(donorIndexA);
+            groupsOfIndices[groupsOfIndicesIndexB].Remove(donorIndexB);
+        }
+
+
+        public IEnumerable<ChromatographicPeakData> CreateChromatographicPeakData(List<DonorGroup> donors, List<int> donorIndices, int maxThreads)
```
Comment (trishorts) [inline mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs:404]: method comment

File: mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs (hunk 20 above)
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/PEP/PepAnalysisEngine.cs:234]: yes

File: mzLib/FlashLFQ/FlashLFQResults.cs (hunk 16 above)
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/FlashLFQResults.cs:72]: For MergeResults? I didn't write it and I don't know what it does

**Review by nbollis — DISMISSED** [review]
I concur with shortreeds comments about comments

File: mzLib/FlashLFQ/FlashLfqEngine.cs (hunk 17 above)
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/FlashLfqEngine.cs:1101]: Not directly, no. A lot of this coed existed previously but was refactored to make it more modular

File: mzLib/FlashLFQ/SpectraFileInfo.cs [hunk 24]
```diff
@@ -39,5 +41,9 @@ public override int GetHashCode()
         {
             return FullFilePathWithExtension.GetHashCode();
         }
+        public override string ToString()
+        {
+            return Path.GetFileName(FullFilePathWithExtension);
+        }
```
[the remark this answers is not in the document]
Comment (Alexander-Sol) [reply mzLib/FlashLFQ/SpectraFileInfo.cs:47]: The method is never actually called, so I'm not too worried

**Comment by Alexander-Sol** [comment]
image
image

**Review by trishorts — DISMISSED** [review]

File: mzLib/mzLib.nuspec [hunk 25]
```diff
@@ -2,7 +2,7 @@
 <package>
   <metadata>
     <id>mzLib</id>
-    <version>1.0.547</version>
+    <version>7.7.72</version>
     <title>mzLib</title>
```
Comment (nbollis) [inline mzLib/mzLib.nuspec:6]: Ope

**Review by nbollis — CHANGES_REQUESTED** [review]
Fix the nuspec and you got my check mark

**Review by trishorts — DISMISSED** [review]

**Review by RayMSMS — DISMISSED** [review]

**Review by trishorts — APPROVED** [review]

**Comment by Alexander-Sol** [comment]
| Comparison   | v1.0 Avg FC | PIP-ECHO Avg. FC | Empirical FC
| :---------------- | ------: | ----: |------: | 
| 1x vs 1.5x         | 0.484 | 0.497 | 0.585 | 
| 1x vs 2x           | 0.857 | 0.871 | 1 | 
|1x vs 2.5x         | 1.16 | 1.16 | 1.32 | 
| 1x vs 3x           |  1.48 | 1.48 | 1.58 | 


| Comparison   |  v1.0 CV | PIP-ECHO CV | 
| :---------------- | ----: |------: | 
| 1x vs 1.5x         |  0.854 | 0.619 | 
| 1x vs 2x           |  0.544 | 0.405 | 
|1x vs 2.5x         | 0.436 | 0.324 | 
| 1x vs 3x           |  0.370 | 0.276 |

**Review by trishorts — APPROVED** [review]

**Review by RayMSMS — APPROVED** [review]

**Review by nbollis — APPROVED** [review]

**Pull request merged**